### PR TITLE
Optimize user-client session relationship for HotRod storage

### LINKED ISSUE #12818: Optimize UserSession->ClientSession relationship for HotRod
#12817 introduces a performance issue for HotRod, this should be solved by splitting user sessions and client sessions into two places so it is possible to update them separately. See https://github.com/keycloak/keycloak/discussions/12788 for more details

### PR DESCRIPTION
**Pull request by mhajas**
Closes #12818

### CODE REVIEW

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        s.realms().removeRealm(realmId);
+    }
+
+    @Override
+    protected boolean isUseSameKeycloakSessionFactoryForAllThreads() {
+        return true;
+    }
+
+    @Test
+    public void testConcurrentNotesChange() {
```
Comment (martin-kanis) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java:72]: What exactly is being tested here? When this test can fail and what determines a successful run?

**Review by martin-kanis — CHANGES_REQUESTED** [review]
Thanks @mhajas  for the PR. I like the way how you implemented it. Please, just see one nitpick and one question.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java (hunk 1 above)
Comment (mhajas) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java:72]: During development, I was getting some `ConcurrentModificationException` so I created this test to reproduce. This is doing a similar thing as `TokenManager` does [here](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L539-L564). 

Maybe we can add some assertions like there should be some amount of clients session and each has some value of the note. I will add it.

**Review by martin-kanis — DISMISSED** [review]
Thanks @mhajas!

**Review by hmlnarik — DISMISSED** [review]
Approving per @martin-kanis's approval

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/HotRodUserSessionClientSessionRelationshipTest.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    @Test
+    public void testSessionsAreRemovedOnUserRemoval() {
+        AtomicReference<String> uSessionId = new AtomicReference<>();
+        AtomicReference<String> cSessionId = new AtomicReference<>();
+        prepareSessions(uSessionId, cSessionId);
+
+        withRealm(realmId, (session, realm) -> {
+            session.clients().removeClient(realm, CLIENT0_CLIENT_ID);
```
Comment (martin-kanis) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/session/HotRodUserSessionClientSessionRelationshipTest.java:92]: It is also necessary to remove the client here?

**Review by martin-kanis — COMMENTED** [review]
@mhajas Thank you for the investigation of the issue and the fix. It looks good, I have just one question.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/HotRodUserSessionClientSessionRelationshipTest.java (hunk 2 above)
Comment (mhajas) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/HotRodUserSessionClientSessionRelationshipTest.java]: No, it is not. It is a relict from a similar test I created for clients. Then I realised we are not removing client sessions on client removal. We are just checking whether the client exists on querying client sessions.

**Review by martin-kanis — APPROVED** [review]
Thanks Michal, failures seem not related to the PR.

**Review by hmlnarik — APPROVED** [review]
Approving per @martin-kanis 's review

**Pull request merged**